### PR TITLE
Dictionary: include the key in the "duplicate key" exception message

### PR DESCRIPTION
**Pull request by akoeplinger**
Fixes https://github.com/dotnet/corefx/issues/1187
Related fix for ImmutableCollections: https://github.com/dotnet/corefx/pull/185

/cc @ellismg

### CODE REVIEW

**Comment by OtherCrashOverride** [comment]
I commented in the corefx thread about why this is undesirable.  As as alternative, how about adding a new field for these extended exception messages that developers may access.  This would leave the current exception messages to source from digitally singed assemblies of known content while giving access to additional information for those that want it:

```
System.Exception.Message
System.Exception.ExtendedMessage
```

**Comment by ayende** [comment]
That would be a pretty bad idea.
You can see how much problem this cause with ReflectionTypeLoadException
and the LoaderException model.
It doesn't get logged most cases, and it requires special handling
everywhere to get it working properly to actually output the full info.

Exception are exposed in very few cases:
* Debug mode - when I want to read the exceptions full info so I can fix
stuff.
* Logs - in which case full information is always best.

You don't show exception information to the customer directly, so that
doesn't matter.
And having the full information there is hugely important in fixing errors.
Otherwise you get a bug report, have no idea what is going on, then have to
send an update that just add extra logging to figure it out.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by OtherCrashOverride** [comment]
> And having the full information there is hugely important in fixing errors.

These exception messages don't give us the full information.  They are of limited use to a limited subset of real world problems.  This is discussed in the corefx thread.

The actual feature wanted is a core dump that someone can attach a remote Visual Studio debug session to and see the EXACT state the program was in when it crashed.  Why are we still "printf debugging"?

**Comment by ayende** [comment]
@OtherCrashOverride

We can't get a crash dump from our customers. They use our software to
manage things like military installations, health care records, financial
transactions, etc.
While a crash dump is nice, it require quite a bit of expertise to use, and
as I said, in many cases, utterly impossible to get.
It also doesn't help if this particular exception happens 50,000 times a
second.

A log file, on the other hand, can either be sent, or the relevant portions
with the full error can be sent, much more easily.
It allows you to monitor distributed systems easily and quickly.

And getting the full error means that you don't need a debugging expert to
fix a bug.


*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by OtherCrashOverride** [comment]
> We can't get a crash dump from our customers. They use our software to manage things like military installations, health care records, financial transactions, etc.

And how do your customers feel about the runtime logging and reporting this sensitive information arbitrarily in exceptions?

**Comment by ayende** [comment]
What kind of sensitive information do you think are actually going on in
exceptions?

Beside, that goes into a file / system that _they_ control. It isn't
exposed arbitrarily.
It also means that if they need to send me a file, they can go through it
very quickly.

About the most sensitive thing that I saw in exception was something like
"ProjectManhattanDesignDocs.docx is not accessible", and that is something
that they can easily filter.


*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by OtherCrashOverride** [comment]
You should be able to make that decision for your company and your customers.  I however should also have the same option.  The information reported is useless ["Duplicate Key: System.Object" doesn't tell me anything].  It also widens the attack surface [as noted in the other thread].  I need to ensure that these types of exception messages are NEVER generated.

I hope a compromise presents.

**Comment by Eyas** [comment]
Why is it not viable to only generate this message in `DEBUG`? Or was this simply never discussed?

(i.e. change `FEATURE_CORECLR` to `FEATURE_CORECLR && DEBUG`)

**Comment by ayende** [comment]
Because that doesn't help when the error happens in production.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by ayende** [comment]
How often do you think people just put plain Object as a dictionary key?
In most cases, it is something that has a sensible ToString() impl.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by Eyas** [comment]
But in production we often lose a lot of debugging ability anyway; debug symbols are missing, the generated code is optimized and more difficult to traverse, crash dumps are less useful, or, in your case, customers are not willing to share them anyway. A better message seems like it could save some debugging time, so at a minimum, it is **definitely** useful with `DEBUG` on.

You're also saying its useful without `DEBUG` on. Sure. But some people are raising 'security' concerns.

I don't know what the outcome of the debate will be. But if the conclusion is that security concerns outweigh the gain in productivity, then I hope the resulting decision is to at least still implement this in, say, `DEBUG`.

**Comment by stephentoub** [comment]
> then I hope the resulting decision is to at least still implement this in, say, DEBUG.

Just to be clear, using a DEBUG compilation constant in mscorlib would only help if you were debugging your application using a debug build of CoreCLR and mscorlib; it wouldn't matter whether your application was a debug or release build.

**Comment by ayende** [comment]
You can deploy with PDB, that is a pretty standard behavior, since stack
traces are much better then.
Sure, you might get methods optimized away, but it is still clear.

If they have security concerns, don't show exceptions to users. You
shouldn't anyway if this is something that concerns you.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by Eyas** [comment]
@stephentoub,

> Just to be clear, using a DEBUG compilation constant in mscorlib would only help if you were debugging your application using a debug build of CoreCLR and mscorlib

bleh, you're right. Compile-time checks wouldn't work. I wonder if there's a runtime check that would achieve the effect I'm talking about. I assume we don't care too much about performance when throwing exceptions (a branch/variable check seems insignificant compared to throwing), but maybe we do.

**Comment by akoeplinger** [comment]
@Eyas making this debug-only would severely reduce the usability of this change. As @ayende notes, it's often impossible/hard to reproduce certain failures during debugging and log files are your only hope of understanding what went wrong. Every little bit of additional information helps.

**Comment by masonwheeler** [comment]
I've seen this line of "reasoning" before.  @OtherCrashOverride does not appear to understand the difference between server software, in which the software is run by the developer and the end-user can be an attacker, and client-side software, in which the software is already in the possession of the end-user, who is therefore, by definition, not an attacker.

His objections *almost* make sense in the server-side world... except that a trivial solution exists: *just don't show error messages to the end-user*.  Because server-side software is run by developers, not end-users, we can dump all necessary error messages to locations where we can see the bugs and we don't have to show them to the users.

But client-side software is a completely different beast.  The software is under the control of someone other than the developers, which means that when something goes wrong, the person who has possession of the software (usually) has no way to fix it, and the people who have the means to fix it do not have possession of the instance of the software that is erroring out.  This makes traditional debugging techniques impossible, which makes it imperative that bug reports contain as much relevant information as possible, in order for the developers to reproduce the problem and be able to fix it.

This, specifically, is the use case, the problem that this PR solves, and it's definitely worth solving.  In the server-side cases that OtherCrashOverride is worried about, again, the solution is trivial: just don't show the user your exception messages in the first place.  But even if you're not showing it to the user, you still want that information for yourself.  So yes, let's merge this.

**Comment by ayende** [comment]
@masonwheeler
Note that in both cases, for client side & server side, having this
additional information is important.

*Hibernating Rhinos Ltd  *

Oren Eini* l CEO l *Mobile: + 972-52-548-6969

Office: +972-4-622-7811 *l *Fax: +972-153-4-622-7811

**Comment by OtherCrashOverride** [comment]
>  @OtherCrashOverride does not appear to understand the difference between server software, in which the software is run by the developer and the end-user can be an attacker, and client-side software, in which the software is already in the possession of the end-user, who is therefore, by definition, not an attacker.

I know him personally and I can tell you for a fact that @OtherCrashOverride does understand a great deal about a great many things; server and client-side software included!

**Comment by sharwell** [comment]
I remain against this proposal for the following reasons:

1. `key.ToString()` is occasionally a meaningful representation of the data used for determining key equality, but it is also frequently meaningless. Including it for the latter cases would be misleading.
1. `key.ToString()` may be a expensive operation (calculation cost or produce a very large string in memory). While each of these may be transient at runtime, the latter could cause unexpected flooding of existing logging databases.
1. I continue to question the security ramifications of this change, and do not believe the benefits outweigh the risk.

I would likely be in support of an alternate proposal that added a read-only `Key` or `DuplicateKey` property to the exception object itself (property type `object`), provided the value was not included in the result of `Message` or `ToString()` by default.

**Comment by OtherCrashOverride** [comment]
> I would likely be in support of an alternate proposal that makes the duplicate key available through some other means, provided the value was not included in the result of Message or ToString() by default.

This is being discussed in dotnet/corefx#3025

File: src/mscorlib/src/System/ThrowHelper.cs [hunk 1]
```diff
@@ -53,6 +53,12 @@ internal static class ThrowHelper {
             throw new ArgumentException(Environment.GetResourceString("Arg_WrongType", value, targetType), "value");
         }
 
+#if FEATURE_CORECLR
+        internal static void ThrowAddingDuplicateWithKeyArgumentException(object key) {
+            throw new ArgumentException(Environment.GetResourceString("Argument_AddingDuplicateWithKey", key));
```
Comment (stephentoub) [inline src/mscorlib/src/System/ThrowHelper.cs]: What happens if key.ToString() throws?

Comment (akoeplinger) [inline src/mscorlib/src/System/ThrowHelper.cs]: That's evil :laughing: 

See also @nguerrera's comment in https://github.com/dotnet/corefx/pull/185#issuecomment-65754562

**Comment by weshaggard** [comment]
@akoeplinger thanks for fixing this and sorry it has been sitting here for so long. The change LGTM.

**Comment by akoeplinger** [comment]
:smile: This is a pretty awesome christmas present, thanks @weshaggard ! :christmas_tree: :tada: :gift:

**Pull request merged**